### PR TITLE
support cuda for vs generator

### PR DESCRIPTION
**Pull request by xq114**
https://github.com/xmake-io/xmake/issues/2077

给vs generator 加上了cuda支持

### CODE REVIEW

**Comment by xq114** [comment]
https://github.com/xmake-io/xmake/blob/master/xmake/plugins/project/vstudio/impl/vs201x.lua#L66-L90
https://github.com/xmake-io/xmake/blob/master/xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua#L31-L41

为啥这里有些escape，有些不escape？

**Comment by waruqi** [comment]
> https://github.com/xmake-io/xmake/blob/master/xmake/plugins/project/vstudio/impl/vs201x.lua#L66-L90 https://github.com/xmake-io/xmake/blob/master/xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua#L31-L41
> 
> 为啥这里有些escape，有些不escape？

因为不是一个人贡献的，第一个是人家贡献进来的，下面那个是我之前搞得，暂时没躺到相关坑，无法验证可用性，所以没去同步，等后面如果如果问题了，再测试同步

File: xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua [hunk 1]
```diff
@@ -402,24 +494,162 @@ function _make_source_options(vcxprojfile, flags, condition)
...
+                function (def)
+                    defstr = defstr .. _escape(def) .. ";"
+                end
+            )
+        end
+        defstr = defstr .. "%%(Defines)"
+        vcxprojfile:print("<Defines%s>%s</Defines>", condition, defstr)
+
+        -- make Include
+        if flagstr:find("%-I") then
```
Comment (waruqi) [inline xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua:554]: “[%-/]I” 更加通用点

File: xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua [hunk 2]
```diff
@@ -402,24 +494,162 @@ function _make_source_options(vcxprojfile, flags, condition)
...
+            local wlevel = flagstr:find("[%-/](W[1234])")
+            vcxprojfile:print("<Warning%s>%s</Warning>", condition, wlevel)
+        elseif flagstr:find("[%-/]Wall") then
+            vcxprojfile:print("<Warning%s>Wall</Warning>", condition)
+        end
+
+        -- make Defines
+        local defstr = ""
+        for _, flag in ipairs(flags) do
+            flag:gsub("%-D(.*)",
```
Comment (waruqi) [inline xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua:544]: 还有这

File: xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua [hunk 3]
```diff
@@ -402,24 +494,162 @@ function _make_source_options(vcxprojfile, flags, condition)
...
+            )
+        end
+        defstr = defstr .. "%%(Defines)"
+        vcxprojfile:print("<Defines%s>%s</Defines>", condition, defstr)
+
+        -- make Include
+        if flagstr:find("%-I") then
+            local dirs = {}
+            for _, flag in ipairs(flags) do
+                flag:gsub("%-I(.*)", function (dir) table.insert(dirs, _escape(dir)) end)
```
Comment (waruqi) [inline xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua:557]: 这里

File: xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua [hunk 4]
```diff
@@ -26,8 +26,34 @@ import("core.project.project")
...
+    ,    ["*"] = "%2A" -- Wildcard character for use in file names in Include and Exclude attributes
+    -- html entities
+    ,    ["\""] = "&quot;"
+    ,    ["<"] = "&lt;"
+    ,    [">"] = "&gt;"
+    ,    ["&"] = "&amp;"
+    }
+
+    return (string.gsub(str, "[%%%$@';%?%*\"<>&]", function (c) return assert(map[c]) end))
+end
```
Comment (waruqi) [inline xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua:55]: 这里既然也已经同步了，那就抽取到独立文件 import 复用下好了，可以放到单独的 vsutils.lua 里面去，作为 vsutils.escape 辅助接口

**Comment by waruqi** [comment]
@SirLynix @yamashi This patch has changed a lot. I'm not sure if it will have other effects. Can you help me test it together? Thanks!

**Comment by SirLynix** [comment]
Sure, how can we update xmake to this PR for testing?

**Comment by xq114** [comment]
> Sure, how can we update xmake to this PR for testing?

xmake update -s github:xq114/xmake#dev

File: xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua [hunk 5]
```diff
@@ -386,10 +462,10 @@ function _make_source_options(vcxprojfile, flags, condition)
     end
 
     -- make AdditionalIncludeDirectories
-    if flagstr:find("[%-/]I") then
+    if flagstr:find("^[%-/]I") then
```
Comment (SirLynix) [inline xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua:465]: This breaks AdditionalIncludeDirectories

File: xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua [hunk 6]
```diff
@@ -402,24 +478,192 @@ function _make_source_options(vcxprojfile, flags, condition)
...
+                function (def)
+                    defstr = defstr .. vsutils.escape(def) .. ";"
+                end
+            )
+        end
+        defstr = defstr .. "%%(Defines)"
+        vcxprojfile:print("<Defines%s>%s</Defines>", condition, defstr)
+
+        -- make Include
+        if flagstr:find("^[%-/]I") then
```
Comment (SirLynix) [inline xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua:538]: Same here

File: xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua [hunk 7]
```diff
@@ -422,7 +422,7 @@ function _make_source_options_cl(vcxprojfile, flags, condition)
     -- make PreprocessorDefinitions
     local defstr = ""
     for _, flag in ipairs(flags) do
-        flag:gsub("^[%-/]D(.*)",
+        flag:gsub("[%-/]D(.*)",
```
Comment (SirLynix) [inline xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua:425]: It was fine here, only problematic on full flag str (`flagstr`)

File: xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua [hunk 8]
```diff
@@ -462,10 +462,10 @@ function _make_source_options_cl(vcxprojfile, flags, condition)
     end
 
     -- make AdditionalIncludeDirectories
-    if flagstr:find("^[%-/]I") then
+    if flagstr:find("[%-/]I") then
         local dirs = {}
         for _, flag in ipairs(flags) do
-            flag:gsub("^[%-/]I(.*)", function (dir) table.insert(dirs, vsutils.escape(dir)) end)
+            flag:gsub("[%-/]I(.*)", function (dir) table.insert(dirs, vsutils.escape(dir)) end)
```
Comment (SirLynix) [inline xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua:468]: and here

File: xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua [hunk 9]
```diff
@@ -525,7 +525,7 @@ function _make_source_options_cuda(vcxprojfile, flags, opt)
         -- make Defines
         local defstr = ""
         for _, flag in ipairs(flags) do
-            flag:gsub("^[%-/]D(.*)",
+            flag:gsub("[%-/]D(.*)",
```
Comment (SirLynix) [inline xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua:528]: and here

File: xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua [hunk 10]
```diff
@@ -535,10 +535,10 @@ function _make_source_options_cuda(vcxprojfile, flags, opt)
         vcxprojfile:print("<Defines%s>%s</Defines>", condition, defstr)
 
         -- make Include
-        if flagstr:find("^[%-/]I") then
+        if flagstr:find("[%-/]I") then
             local dirs = {}
             for _, flag in ipairs(flags) do
-                flag:gsub("^[%-/]I(.*)", function (dir) table.insert(dirs, vsutils.escape(dir)) end)
+                flag:gsub("[%-/]I(.*)", function (dir) table.insert(dirs, vsutils.escape(dir)) end)
```
Comment (SirLynix) [inline xmake/plugins/project/vstudio/impl/vs201x_vcxproj.lua:541]: and here

**Review by SirLynix — COMMENTED** [review]
otherwise, it seems good

**Comment by waruqi** [comment]
@SirLynix Does it work for you now? If there are no other issues, I am going to merge this patch.

**Comment by SirLynix** [comment]
Seems good 👌

**Pull request merged**